source/request_source/request_source_plugin_impl.h

```diff
@@ -0,0 +1,124 @@
+// Stub Request Source implementation for comparison.
+class DummyRequestSource : public RequestSource {
+public:
+  explicit DummyRequestSource(const nighthawk::request_source::StubPluginConfig& config);
+  // The generator function will only return empty headers.
+  // The function is threadsafe.
+  RequestGenerator get() override;
+
+  // default implementation
+  void initOnThread() override;
+};
+
+// Factory that creates a DummyRequestSource from a DummyRequestSourcePluginConfig proto.
+// Registered as an Envoy plugin.
+// Stub implementation of RequestSourceConfigFactory which produces a RequestSource.
+// RequestSources are used to get RequestGenerators which generate requests for the benchmark
+// client. All plugins configuration are specified in the request_source_plugin.proto This class is
+// thread-safe, but it doesn't do anything. Usage: assume you are passed an appropriate Any type
+// object called config, an Api object called api, and a default header called header. auto&
+// config_factory =
+//     Envoy::Config::Utility::getAndCheckFactoryByName<RequestSourcePluginConfigFactory>(
+//         "nighthawk.stub-request-source-plugin");
+// RequestSourcePtr plugin =
+//     config_factory.createRequestSourcePlugin(config, std::move(api), std::move(header));
+
+class DummyRequestSourcePluginConfigFactory : public virtual RequestSourcePluginConfigFactory {
+public:
+  // This is a hardcoded string.
+  std::string name() const override;
+  // This returns an empty version of the expected StubPluginConfig from request_source_plugin.proto
+  Envoy::ProtobufTypes::MessagePtr createEmptyConfigProto() override;
+  // This is the primary method that is used to get RequestSources.
+  // This implementation is thread safe, but the RequestSource it generates doesn't do much.
+  RequestSourcePtr createRequestSourcePlugin(const Envoy::Protobuf::Message& message,
+                                             Envoy::Api::ApiPtr api,
+                                             Envoy::Http::RequestHeaderMapPtr header) override;
+};
+
+// This factory will be activated through RequestSourceFactory in factories.h
+DECLARE_FACTORY(DummyRequestSourcePluginConfigFactory);
+
+// Sample Request Source for small RequestOptionsLists. Loads a copy of the RequestOptionsList in
+// memory and replays them.
+// @param request_max The number of requests the requestGenerator produced by get() will generate. 0
+// means it is unlimited.
+// @param header the default header that will be overridden by values taken from the options_list,
+// any values not overridden will be used.
+// @param options_list A copy of the options_list will be loaded in memory. The RequestGenerator
+// produced by get() will use options from the options_list to overwrite values in the header, and
+// create new requests. if request_max is greater than the length of options_list, it will loop.
+// This is not thread safe.
+class RequestOptionsListRequestSource : public RequestSource {
```

Comment:
What is the practical use case for this request source implementation? Looking back, our plan was to implement the file based request source to serve as an example and a gRPC request source for backward compatibility.

Can you help me understand what do we need this one for?

EDIT: After looking at the .cc file, it looks like this request source implementation is used internally by the FileBased request source to do its job. If that it the plan, that is fine, but then this shouldn't be part of the API. The API (the header) should only contain the FileBased request source. The .cc file can have a section with anonymous namespace where we define the RequestOptionsListRequestSource internally.

Alternatively, if we plan to reuse this by multiple request sources, we should move it into its own library and say (in the header file) that this is for internal purposes only.

Reply:
I felt like it made sense for both the RequestSource and the factory that makes the requestSource to live in the same file together, I can move it into its own library if you prefer.